src/queue.c

```diff
@@ -0,0 +1,276 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_config.h>
+#include <packetgraph/errors.h>
+#include <packetgraph/queue.h>
+#include "utils/bitmask.h"
+#include "brick.h"
+#include "packets.h"
+
+struct pg_queue_config {
+	int rx_max_size;
+};
+
+struct pg_queue_state {
+	struct pg_brick brick;
+	/* side where packets are burst */
+	enum pg_side output;
+	/* store packets bursted in queue */
+	int rx_max_size;
+	GAsyncQueue *rx;
+	/* Queue's friend */
+	struct pg_queue_state *friend;
+};
+
+struct pg_queue_burst {
+	struct rte_mbuf **pkts;
+	uint64_t mask;
+};
+
+static struct pg_brick_config *queue_config_new(const char *name,
+						int rx_max_size)
+{
+	struct pg_brick_config *config = g_new0(struct pg_brick_config, 1);
+	struct pg_queue_config *queue_config = g_new0(struct pg_queue_config,
+						      1);
+
+	queue_config->rx_max_size = rx_max_size;
+	config->brick_config = (void *) queue_config;
+	return pg_brick_config_init(config, name, 1, 1, PG_MONOPOLE);
+}
+
+uint8_t pg_queue_pressure(struct pg_brick *queue)
+{
+	struct pg_queue_state *state =
+		pg_brick_get_state(queue, struct pg_queue_state);
+	int queue_size = g_async_queue_length(state->rx);
+
+	return queue_size <= 0 ? 0 : queue_size * 255 / state->rx_max_size;
+}
+
+/* the fastpath data function of the queue_brick just forward the bursts */
+static int queue_burst(struct pg_brick *brick, enum pg_side from,
+		       uint16_t edge_index, struct rte_mbuf **pkts,
+		       uint64_t pkts_mask, struct pg_error **error)
+{
+	struct pg_queue_state *state =
+		pg_brick_get_state(brick, struct pg_queue_state);
+	struct pg_queue_burst *burst = NULL;
+
+	/* the oldest burst is throw away */
+	if (g_async_queue_length(state->rx) >= state->rx_max_size) {
+		burst = g_async_queue_try_pop(state->rx);
+		if (likely(burst != NULL)) {
+			pg_packets_free(burst->pkts, burst->mask);
+			g_free(burst);
+		}
+	}
+
+	burst = g_new(struct pg_queue_burst, 1);
+	burst->pkts = pkts;
+	burst->mask = pkts_mask;
+	g_async_queue_push(state->rx, burst);
+
+	return 0;
+}
+
+static int queue_poll(struct pg_brick *brick, uint16_t *pkts_cnt,
+		      struct pg_error **error)
+{
+	uint16_t nb_pkts;
+	int ret = 0;
+	GAsyncQueue *tx = NULL;
+	struct pg_queue_burst *burst = NULL;
+	struct pg_queue_state *state =
+		pg_brick_get_state(brick, struct pg_queue_state);
+	struct pg_brick_side *s = &brick->side;
+
+	/* Do we have a queue friend where to poll packets ? */
+	if (!state->friend) {
+		*pkts_cnt = 0;
+		return 0;
+	}
+
+	tx = state->friend->rx;
+	burst = g_async_queue_try_pop(tx);
+	if (!burst) {
+		*pkts_cnt = 0;
+		return 0;
+	}
+
+	nb_pkts = pg_mask_count(burst->mask);
+	*pkts_cnt = nb_pkts;
+
+	if (unlikely(s->edge.link == NULL))
```

Comment:
maybe we should check this before polling the packets from the queue.
this memory leak the burst, if s->edge.link is NULL BTW.